main/src/com/google/refine/grel/controls/ForRange.java

```diff
@@ -83,32 +83,46 @@ public Object call(Properties bindings, Evaluable[] args) {
                 long step = ((Number) stepO).longValue();
                 double to = ((Number) toO).doubleValue();
 
-                if (!((from < to && step > 0) || (from > to && step < 0))) {
+                if (step == 0) {
                     return new EvalError(EvalErrorMessage.invalid_arg());
                 }
 
-                int rangeSize = (int) Math.ceil(Math.abs(from - to) / Math.abs(step));
-                for (int i = 0; i < rangeSize; i++) {
-                    bindings.put(indexName, from);
-                    Object r = args[4].evaluate(bindings);
-                    results.add(r);
-                    from += step;
+                if (step <= Math.abs(from - to)) {
+                    while (from < to && step > 0) {
```

Comment:
`step` is invariant here, so I would either convert this an `if` statement
```
if (step > 0) {
  while (from < to) {
  }
}
else {
  while (from < to) {
  }
}
```
or, to reduce the redundancy, fold both loops together and make the conditional slightly more complex
```
while ((step > 0 && from < to) || to > from) {
```
(we don't need to test `step < 0` because we know it's not zero and not greater than zero)